Using datasets for image search blog post

davanstrien opened the pull request:
cc @osanseviero 

Thanks again for reaching out. I've created an initial draft of a blog post on using 🤗 `datasets` for image search based on: https://danielvanstrien.xyz/metadata/deployment/huggingface/ethics/huggingface-datasets/faiss/2022/01/13/image_search.html

I've made some edits already, mainly to fix a few typos and improve the clarity in places. I have a couple of changes I wanted to pass by you and a few questions. Happy to hear any other comments/suggestions 😀

## Things I still need to do
- [x] Upload images associated with the post (see below for a question on how many images makes sense). 
- [x] Make suggested changes (see below)
- [x] Finalise Colab notebook (see below)

## Suggested changes

- There is quite a bit of discussion in the blog post, which is focused on loading images in `pillow` without the `fname` attribute. I did this to make sure `push_to_hub` worked as expected. Since https://github.com/huggingface/datasets/pull/3685 was merged, this is no longer necessary, and `datasets` takes care of this for you. I would suggest cutting the 'Where is the image 🤔' section to reflect this change in the `datasets` library. @osanseviero  - if that sounds good to you, I can edit that part out. 

## Questions:

- At the moment, there are quite a lot of images in the original blog post, and I'm not sure if it makes sense to include all of these or not? I was thinking of maybe only including two search output examples in the blog post and keeping all the outputs in a related colab notebook. Does that sound okay to you?
- Is there a convention in the blog posts of assuming people are in a notebook environment? So far, I've switched to assuming a normal bash environment for commands i.e. `pip install` rather than the notebook version `!pip install`. Let me know if you prefer one or the other. 
- I thought it would be nice to include an accompanying Colab notebook:
	- 	For this, I would suggest using a sample of the dataset as the default since the dataset I used in the original post is relatively large, and the downloading and unzipping takes up a good chunk of time - does that sound okay to you? 
	- Is there somewhere specific I should host this notebook on GitHub, or is a personal repository okay?

NielsRogge commented:
Hi, thanks for your PR (and the awesome blog post you made).

I'd like to note that any assets (like images such as 2022-01-13-image_search_files/2022-01-13-image_search_19_0.png) can be placed in the [assets directory](https://github.com/huggingface/blog/tree/master/assets). This directory can also contain a nice thumbnail, to be shown when sharing the post on social media.

You'll also need to add the blog to the `_blog.yml` file. I'd recommend to check out a previous PR (such as #227) as an illustration.

osanseviero commented:
Thanks again for the PR! :rocket: Sorry for the delay

Following up Niels replies and your questions

> At the moment, there are quite a lot of images in the original blog post, and I'm not sure if it makes sense to include all of these or not? I was thinking of maybe only including two search output examples in the blog post and keeping all the outputs in a related colab notebook. Does that sound okay to you?

If you want to do that, it's ok. It's also ok if you want to add more images, we do try to compress them so they don't weight too much in the repo.

> Is there a convention in the blog posts of assuming people are in a notebook environment? So far, I've switched to assuming a normal bash environment for commands i.e. pip install rather than the notebook version !pip install. Let me know if you prefer one or the other.

No, we don't have a convention. We usually add a link to a colab at the top for this kind of blog posts. https://huggingface.co/blog/fine-tune-vit

> For this, I would suggest using a sample of the dataset as the default since the dataset I used in the original post is relatively large, and the downloading and unzipping takes up a good chunk of time - does that sound okay to you?

Using a sample makes sense to me.

> Is there somewhere specific I should host this notebook on GitHub, or is a personal repository okay?

A personal repository is good!

davanstrien commented:
> This is awesome ! I just added a comment about the part about `push_to_hub` to bundle the images inside the parquet files :)

Thanks! I've just updated the post to reflect this change in `datasets`. 

> If it makes sense we can do a release of `datasets` soon and update the code, since the workaround you mentioned is not going to be necessary anymore

That would be amazing. At the moment in the Colab I'm install from git - I can change that when there is a new `datasets` release.

davanstrien commented:
@NielsRogge @osanseviero @lhoestq, thank you all for your comments and suggestions -  I think I've addressed most of these now.  

I have pushed the images; so far, I have just converted them to jpegs but let me know if you have other requirements for image compression. I've also added an entry to the YAML file. 

I haven't added a header to the blog post yet. Should I modify [this](https://github.com/huggingface/blog/blob/e25516f13f4a1a62eff2ea3ab0338cc8d5c3c545/fine-tune-vit.md?plain=1#L10-L25) for my post? Hopefully, other than that, the blog post is ready to go but let me know if you have other suggestions.

davanstrien commented on image-search-datasets.md:
```diff
@@ -0,0 +1,739 @@
+
+One of the super awesome things about the huggingface ecosystem is the huggingface hub. We can use the hub to access models and datasets. Often this is used for sharing work with others but it can also be a useful tool for work in progress. The datasets library recently added a `push_to_hub` method that allows you to push a dataset to the hub with minimal fuss. This can be really helpful by allow ...
+
+When I started playing around with this feature I was also keen to see if it could be used as a way of 'bundling' everything together. This is where I noticed that if you push a dataset containing images which have been loaded in from filepaths by pillow the version on the hub won't have the images attached. If you always have the image files in the same place when you work with the dataset then  ...
```
Very excited about this feature, it's already been really useful for another project I'm working on.

nateraw commented:
@davanstrien yes, you can follow the process in that PR to add the headers! Just copy paste and replace with your information. I'll also give this a review later on today or early tomorrow 😄

mariosasko reviewed (commented):
Impressive!

It would be amazing to showcase the new ImageFolder dataset (PR [here](https://github.com/huggingface/datasets/pull/2830/files)) in this blog post. This way, the entire pipeline, from loading to preprocessing, will be cached by default.

This is how loading/preprocessing would look like with the ImageFolder dataset:
```python
from datasets import load_dataset
dataset = load_dataset("imagefolder", data_files="https://zenodo.org/record/6224034/files/embellishments_sample.zip?download=1")
dataset = dataset.rename_column("label", "year") # use map to add fname / path
dataset = dataset.map(img_to_thumbnail) # to save some bandwidth in push_to_hub
```

Let me know if you need help with this.

davanstrien commented:
> @davanstrien yes, you can follow the process in that PR to add the headers! Just copy paste and replace with your information.
will do that first thing tomorrow 
> I'll also give this a review later on today or early tomorrow 😄
thanks 😀

davanstrien commented on image-search-datasets.md:
````diff
@@ -0,0 +1,425 @@
+```
+
+``` python
+dataset = dataset.cast_column('img', features.Image())
````
> It would be amazing to showcase the new ImageFolder dataset (PR [here](https://github.com/huggingface/datasets/pull/2830/files)) in this blog post. This way, the entire pipeline, from loading to preprocessing, will be cached by default.

That sounds great, I've been watching that PR will great anticipation since it will be useful for another project I'm working on 😀

If that PR will be ready to merge soon I'm happy to update the blog to use that. I'll update the notebook/post tomorrow and ping you if I run into any problems. I assume that the public facing API won't change from what is outlined in the current PR?

davanstrien replied:
I have just taken a look at this. Unless I missed something in this case using `ImageFolder` might not work for all the parts in this example. In this dataset the structure follows the imagefolder approach for the most part i.e. 
```
top_folder
	year/
		fname.jpg
```

However, in this case the filename also contains some metadata.

It's possible I missed something but I think the fname isn't preserved when loaded using `ImageFolder`. This [Colab notebook](https://colab.research.google.com/drive/129zbECz8O1UgIm6pjGlcia46EWNya3iW?usp=sharing) shows what I have for this dataset. 

Do you think there is some benefit in preserving the original path as a feature in the dataset when loaded using `ImageFolder`? I think it's not super common to have metadata in the filename but it's also not unkown. This might also make it easier to handle folder which have the structure:

``` 
top_folder
	train/
		dogs/
			dog.1.jpg
		cats/
		pigeon/
	valid/
		dogs/
		cats/
		pigeon/
```

Since you would have access to the path `top_folder/train/dogs/dog1.jpg` and could use that information for splitting in the same way as in the original dataset i.e. some form of `map(lambda x: {"is_train": x['fpath'].split('/')[:-2]=="train"})

davanstrien replied:
I have just realised that you can get this information from the `Image.filename`. i.e. `dataset['train'][0]['image'].filename`.  I can use that for this example to get the other required metadata. I am still wondering if making this a feature would make sense (maybe as a flag) but there might be other downsides I haven't considered.

mariosasko replied:
Yes, the public-facing API should stay the same.

Besides `image.filename`, another option is to use:
```python
dataset = dataset.cast_column("image", Image(decode=False))
dataset[0]["image"]["path"]
```
to access path information. My rationale for not having paths as a separate feature is to avoid storing duplicate data in a dataset, and additionally, it avoids local path information leakage in the `push_to_hub`

And to build splits, you can use globbing to match local files or files on the Hub, e.g.:
```python
from datasets import load_dataset
dataset = load_datataset("imagefolder", data_files={"train": "top_folder/train/**/*.jpg", "valid": "top_folder/valid/**/*.jpg"})
```

davanstrien replied:
Thanks, I've just updated to use `ImageFolder` for loading. Everything worked smoothly for this and it's really nice to be able to cut out a whole bunch of loading code using this 😀 For now, in the Colab I am installing from that branch but I will update that if this pull request is likely to make it into the next `datasets` release.

davanstrien commented:
> @davanstrien yes, you can follow the process in that PR to add the headers! Just copy paste and replace with your information. I'll also give this a review later on today or early tomorrow 😄

I have added this now (hopefully correctly).

osanseviero commented on image-search-datasets.md:
```diff
@@ -0,0 +1,395 @@
+
+## Creating a huggingface space? 🤷🏼 
+
+One obvious next step for this kind of project is to create a hugginface [spaces](https://huggingface.co/spaces) demo. This is what I've done for other [models](https://huggingface.co/spaces/BritishLibraryLabs/British-Library-books-genre-classifier-v2)
```
```suggestion
One obvious next step for this kind of project is to create a Hugging Face [Spaces](https://huggingface.co/spaces/launch) demo. This is what I've done for other [models](https://huggingface.co/spaces/BritishLibraryLabs/British-Library-books-genre-classifier-v2)
```

osanseviero reviewed (approved):
This is awesome :fire: I built locally and the rendering looks great! If possible I would consider compressing a bit some of the larger images, but it's ok if we go ahead with this since it's image heavy.

I left a bunch of super minor nits, but IMO we can go ahead and merge. @nateraw would you like to do a review or should we go ahead?

davanstrien commented:
@osanseviero thanks so much for all of your suggestions. These should all be addressed now. I've also compressed the images a little bit more but let me know if you think further compression would be worth doing.

osanseviero commented:
Thank you very much @davanstrien!! :rocket: this blog post is awesome :smile: Let's see if @nateraw has any further feedback and we can deploy this today if not :rocket: :fire:

davanstrien commented:
> Thank you very much @davanstrien!! 🚀 this blog post is awesome 😄 Let's see if @nateraw has any further feedback and we can deploy this today if not 🚀 🔥

Just spotted that the `imagefolder` loader isn't actually included in the 18.4 `datasets` release (https://github.com/huggingface/datasets/compare/1.18.4...master). I can update the notebook to install from `master` but otherwise, we could wait until `imagefolder` is included in a release to publish?

NielsRogge commented on image-search-datasets.md:
````diff
@@ -0,0 +1,392 @@
+
+Depending on where you are running the code, you may need to login using the `huggingface-cli login` command.
+
+``` python
+huggingface-cli login
+```
````
In a notebook, you can run:

```
from huggingface_hub import notebook_login

notebook_login()
```
which is a bit more user friendly compared to the CLI command. In a terminal, you need to run `huggingface-cli login` indeed.

NielsRogge commented on image-search-datasets.md:
````diff
@@ -0,0 +1,392 @@
+dataset = load_dataset("davanstrien/embellishments-sample", use_auth_token=True)
+```
+## Creating embeddings 🕸 
+We now have a dataset with a bunch of images in it. To begin creating our image search app, we need to embed these images. There are various ways to try and do this, but one possible way is to use the clip models via the `sentence_transformers` library. The [clip model](https://openai.com/blog/clip/) from OpenAI learns a joint representation for both images and text, which is very useful for what ...
````
```suggestion
We now have a dataset with a bunch of images in it. To begin creating our image search app, we need to embed these images. There are various ways to try and do this, but one possible way is to use the CLIP models via the `sentence_transformers` library. The [CLIP model](https://openai.com/blog/clip/) from OpenAI learns a joint representation for both images and text, which is very useful for what we want to do since we want to input text and get back an image.
```

NielsRogge commented on image-search-datasets.md:
````diff
@@ -0,0 +1,392 @@
+We can download the model using the `SentenceTransformer` class.
+
+``` python
+from sentence_transformers import SentenceTransformer, util
+
+model = SentenceTransformer('clip-ViT-B-32')
````
```suggestion
from sentence_transformers import SentenceTransformer

model = SentenceTransformer('clip-ViT-B-32')
```
Not sure whether "util" is needed here.

NielsRogge commented on image-search-datasets.md:
````diff
@@ -0,0 +1,392 @@
+model = SentenceTransformer('clip-ViT-B-32')
+```
+
+This model will take as input either an image or some text and return an embedding. We can use the `datasets` `map` method to encode all our images using this model. When we call map, we return a dictionary with the key `embeddings` containing the embeddings returned by the embedding model. We also pass `device='cuda'` when we call model; this ensures that we're doing the encoding on the GPU.
+
````
```suggestion
This model will take as input either an image or some text and return an embedding. We can use the `datasets` `map` method to encode all our images using this model. When we call map, we return a dictionary with the key `embeddings` containing the embeddings returned by the model. We also pass `device='cuda'` when we call the model; this ensures that we're doing the encoding on the GPU.

```

NielsRogge commented on image-search-datasets.md:
````diff
@@ -0,0 +1,392 @@
+This model will take as input either an image or some text and return an embedding. We can use the `datasets` `map` method to encode all our images using this model. When we call map, we return a dictionary with the key `embeddings` containing the embeddings returned by the embedding model. We also pass `device='cuda'` when we call model; this ensures that we're doing the encoding on the GPU.
+
+``` python
+ds_with_embeddings = dataset.map(
+    lambda example: {'embeddings':model.encode(example['image'], device='cuda')}, batch_size=32)
+```
````
Shouldn't you also set `batched=True`? Since `batched=False` by default.

davanstrien replied:
good catch :)

NielsRogge commented on image-search-datasets.md:
```diff
@@ -0,0 +1,394 @@
+
+🤗 [`datasets`](https://huggingface.co/docs/datasets/) is a library that makes it easy to access and share datasets. It also makes it easy to process data efficiently -- including working with data which doesn't fit into memory.
+
+When `datasets` was first launched it was more usually associated with text data. However, recently, `datasets` has added increased support for images. In particular, there is now a `datasets` [feature type for images](https://huggingface.co/docs/datasets/package_reference/main_classes.html?highlight=image#datasets.Image). A previous [blog post](https://huggingface.co/blog/fine-tune-vit) showed h ...
```
```suggestion
When `datasets` was first launched, it was more usually associated with text data. However, recently, `datasets` has added increased support for images. In particular, there is now a `datasets` [feature type for images](https://huggingface.co/docs/datasets/package_reference/main_classes.html?highlight=image#datasets.Image). A previous [blog post](https://huggingface.co/blog/fine-tune-vit) showed how `datasets` can be used with 🤗 `transformers` to train an image classification model. In this blog post, we'll see how we can combine `datasets` and a few other libraries to create an image search application.
```

NielsRogge commented on image-search-datasets.md:
```diff
@@ -0,0 +1,394 @@
+some nice things 'for free'. One of these is the ability to add a [`faiss`](https://github.com/facebookresearch/faiss) index to a dataset. [`faiss`](https://github.com/facebookresearch/faiss) is a ["library for efficient similarity search and clustering of dense
+vectors"](https://github.com/facebookresearch/faiss).
+
+The `datasets` [docs](https://huggingface.co/docs/datasets) show an [example](https://huggingface.co/docs/datasets/faiss_es.html#id1) of using a `faiss` index for text retrieval. In this post we'll see if we can do the same for images.
```
```suggestion
The `datasets` [docs](https://huggingface.co/docs/datasets) shows an [example](https://huggingface.co/docs/datasets/faiss_es.html#id1) of using a `faiss` index for text retrieval. In this post we'll see if we can do the same for images.
```

NielsRogge commented on image-search-datasets.md:
```diff
@@ -0,0 +1,394 @@
+
+## The dataset: "Digitised Books - Images identified as Embellishments. c. 1510 - c. 1900"
+
+This is a dataset of images which have been pulled from a collection of digitised books from the British Library. These images come from books across a wide time period and from a broad range of domains. The images were extracted using information contained in the OCR output for each book. As a result, it's known which book the images came from but not necessarily anything else about that image i ...
```
```suggestion
This is a dataset of images which have been pulled from a collection of digitised books from the British Library. These images come from books across a wide time period and from a broad range of domains. The images were extracted using information contained in the OCR output for each book. As a result, it's known which book the images came from, but not necessarily anything else about that image i.e. what it is of.
```
That last part is a bit weird to me "what it is of"? What does that mean? :)

NielsRogge commented on image-search-datasets.md:
````diff
@@ -0,0 +1,394 @@
+})
+```
+
+We can get back a DatasetsDict, and we have a Dataset with image and label features. Since we don't have any train/validation splits here, let's grab the train part of our dataset. Let's also take a look at one example from our dataset to see what this looks like.
````
```suggestion
We can get back a `DatasetDict`, and we have a Dataset with image and label features. Since we don't have any train/validation splits here, let's grab the train part of our dataset. Let's also take a look at one example from our dataset to see what this looks like.
```

NielsRogge commented on image-search-datasets.md:
````diff
@@ -0,0 +1,394 @@
+We can use another method from the datasets library `get_nearest_examples` to get images which have an embedding close to our input prompt embedding. We can pass in a number of results we want to get back.
+
+``` python
+scores, retrieved_examples = ds_with_embeddings['train'].get_nearest_examples('embeddings', prompt,k=9)
````
```suggestion
scores, retrieved_examples = ds_with_embeddings['train'].get_nearest_examples('embeddings', prompt, k=9)
```

NielsRogge commented on image-search-datasets.md:
````diff
@@ -0,0 +1,394 @@
+``` python
+def get_image_from_text(text_prompt, number_to_retrieve=9):
+    prompt = model.encode(text_prompt)
+    scores, retrieved_examples = ds_with_embeddings['train'].get_nearest_examples('embeddings', prompt,k=number_to_retrieve)
````
```suggestion
    scores, retrieved_examples = ds_with_embeddings['train'].get_nearest_examples('embeddings', prompt, k=number_to_retrieve)
```

NielsRogge commented on image-search-datasets.md:
```diff
@@ -0,0 +1,394 @@
+
+<img src="assets/53_image_search_datasets/steam_engine_search_results.jpg">
+
+Some of these results look fairly close to our input prompt. We can wrap
+this in a function so can more easily play around with different prompts
```
```suggestion
Some of these results look fairly close to our input prompt. We can wrap
this in a function so we can more easily play around with different prompts
```

NielsRogge commented on image-search-datasets.md:
```diff
@@ -0,0 +1,394 @@
+
+### Trying a bunch of prompts ✨
+
+Now we have a function for getting a few results we can try a bunch of
```
```suggestion
Now we have a function for getting a few results, we can try a bunch of
```

NielsRogge commented on image-search-datasets.md:
```diff
@@ -0,0 +1,394 @@
+
+It was a fairly simple process to get a [Gradio app setup](https://gradio.app/) from the point we got to here. Here is a screenshot of this app:
+
+<img src="assets/53_image_search_datasets/spaces_image_search.jpg" alt="Screnshot of Gradio search app">
```
```suggestion
<img src="assets/53_image_search_datasets/spaces_image_search.jpg" alt="Screenshot of Gradio search app">
```

NielsRogge commented on image-search-datasets.md:
```diff
@@ -0,0 +1,394 @@
+
+> ### Primary intended uses
+>
+> We primarily imagine the model will be used by researchers to better understand robustness, generalization, and other capabilities, biases,  and constraints of computer vision models.
```
```suggestion
> We primarily imagine the model will be used by researchers to better understand robustness, generalization, and other capabilities, biases, and constraints of computer vision models.
```

NielsRogge commented on image-search-datasets.md:
```diff
@@ -0,0 +1,394 @@
+
+- load images into the new `Image` feature type
+- 'save' our work using `push_to_hub` and use this to move data between machines/sessions
+- create a `faiss` index for images that we can use to retrieve images from a text (or image) input
```
```suggestion
- create a `faiss` index for images that we can use to retrieve images from a text (or image) input.
```

julien-c commented:
blogpost looks really cool, @nateraw want to give it a final read? Seems like 100% a topic you'd have good insights on

lhoestq commented on image-search-datasets.md:
````diff
@@ -0,0 +1,394 @@
+We can get back a `DatasetDict`, and we have a Dataset with image and label features. Since we don't have any train/validation splits here, let's grab the train part of our dataset. Let's also take a look at one example from our dataset to see what this looks like.
+
+```python
+dataset[0]
````
I think one line is missing here:
```suggestion
dataset = dataset["train"]
dataset[0]
```

osanseviero commented on _blog.yml:
```diff
@@ -506,4 +506,12 @@
+  title: "Image search with 🤗 datasets"
+  author: davanstrien
+  thumbnail: /blog/assets/53_image_search_datasets/spaces_image_search.jpg
+  date: February 23, 2022
```
Please update :)

osanseviero commented on image-search-datasets.md:
```diff
@@ -0,0 +1,395 @@
+<h1> Image search with 🤗 datasets </h1> 
+
+<div class="blog-metadata">
+    <small>Published March 8, 2022.</small>
```
Please update

osanseviero reviewed (approved):
@davanstrien could you update the two dates? We can merge afterwards and launch this :rocket:

mariosasko commented:
@osanseviero `imagefolder` is only available on the `datasets` master branch, so maybe we should wait for the v2.0 release and have the launch early next week?

davanstrien commented:
> @osanseviero `imagefolder` is only available on the `datasets` master branch, so maybe we should wait for the v2.0 release and have the launch early next week?

Happy to wait until next week for the 2.0 release if that works for you @osanseviero?

osanseviero commented:
Yes, for sure! :rocket: thanks again @davanstrien

osanseviero commented:
Thanks @davanstrien! WDYT of publishing this tomorrow so the communication of this blog post does not collide with `datasets` 2.0.0 announcement? https://twitter.com/qlhoest/status/1503788989299539972 

(We can also do it today if you prefer, but that might be too much datasets for people for a single day)

davanstrien commented:
> Thanks @davanstrien! WDYT of publishing this tomorrow so the communication of this blog post does not collide with `datasets` 2.0.0 announcement? https://twitter.com/qlhoest/status/1503788989299539972

That sounds good too me. I will update dates tomorrow morning.

> (We can also do it today if you prefer, but that might be too much datasets for people for a single day)

Surely too much datasets isn't possible 😜

davanstrien commented:
@osanseviero I've updated the publication date now

osanseviero commented:
Wohooo :rocket:

osanseviero commented:
And deployed!

It seems the thumbnail and some of the images are broken at https://huggingface.co/blog/image-search-datasets and https://huggingface.co/blog

osanseviero commented:
I think it's an issue in our side, looking into it

davanstrien commented:
> I think it's an issue in our side, looking into it

Thanks, let me know if there is something I can do at my end.

osanseviero commented:
It's fully working now! :rocket:

davanstrien commented:
> It's fully working now! 🚀

Awesome!

@NielsRogge @lhoestq @mariosasko @osanseviero thanks so much all for your suggestions/reviews 🤗
